CMakeLists.txt

```diff
@@ -2430,7 +2423,7 @@ set(CPACK_PACKAGE_INSTALL_DIRECTORY "Mixxx")
 set(CPACK_PACKAGE_EXECUTABLES "mixxx;Mixxx")
 set(CPACK_PACKAGE_ICON "${CMAKE_SOURCE_DIR}/res/images/mixxx_install_logo.bmp")
 set(CPACK_PACKAGE_HOMEPAGE_URL "https://www.mixxx.org/")
-set(CPACK_PACKAGE_FILE_NAME "Mixxx-${GIT_BRANCH}-r${GIT_COMMIT_COUNT}-${GIT_COMMIT_HASH}")
+set(CPACK_PACKAGE_FILE_NAME "Mixxx_${GIT_BRANCH}_${GIT_COMMIT_DATE_TIME}_${GIT_COMMIT_HASH}")
```

Comment:
The package name and the Version are separated by an _, the version itself has alphanumerics and +-.~
https://www.debian.org/doc/debian-policy/ch-controlfields.html#version

our debian pachage is currently: 
mixxx_2.2.4~dfsg-1.deb

When we move to a date based revision, we will have 
mixxx_2.3.0-2020-12-24-22-12-59-0ubuntu1.deb 

Fortunately this is considered as more recent. 

This is considered a branch name like: 
mixxx_main-2020-12-24-22-12-59-0ubuntu1 

Will fail to build, because a version number needs to start with a number. 

So we need to use the a version number from 
https://github.com/mixxxdj/mixxx/blob/main/src/_version.h
after the _
2.4.0-alpha-pre-2020-12-24-22-12-59-0ununtu1 

for branches not one of 2.2 2.3 or main we can append the branch name
2.4.0-alpha-pre-2020-12-24-22-12-59-featureX-0ununtu1 

But we need to remove all non alpanumeric digits from the branch name. 

0ununtu1  or whatever needs to be read from an environment variable.

Reply:
I think we should drop the commit hash from the file name. The date and branch is significant enough to pick the right file. 

I also think the version is now quite long. 

Skimming trough the packages, there is no version nearly that long. 

The longest examples are: 
0.0~hg20170929.c53cc1ae4788+dfsg-3
0.0.0+18.04.20171202-0ubuntu1
2.0.1+20172804git0cd03c4f72385-1
0.2.2+git20161105.1.4b157f51-1

Reply:
how about going for 
2.4.0-alpha-pre-feature1-20201224-90b32946ba-0ununtu1
or
2.4.0-alpha-pre-feature1-20201224-90b3-0ununtu1
 

in a release it melds down to
2.4.0-20201224-9083-0ununtu1
or 
2.4.0-20201224-9089-0ununtu1

Drawback: random updates from a PPA during one day 

I think all this makes us originally just count the commits.

Is there a way to count commit per day? 
What are the options to keep the count with GitHub actions?  

.

Reply:
There is no need to count the commits in a day if we include the UTC normalized time of the commit in addition to the date, which this does already.

Reply:
> So we need to use the a version number from
https://github.com/mixxxdj/mixxx/blob/main/src/_version.h
after the _

Then I think we need to move the version string to its own file and generate _version.h with CMake.

Reply:
> But we need to remove all non alpanumeric digits from the branch name.

https://github.com/FranzDiebold/github-env-vars-action can help us here.

Reply:
> I think we should drop the commit hash from the file name. The date and branch is significant enough to pick the right file.

I disagree. It is helpful to be able to look through the files on downloads.mixxx.org and easily know what code the build was made from. I have already had this come in handy today when debugging the macOS code signing stalling.

> I also think the version is now quite long.

What is the problem with that?

Reply:
>  What are the options to keep the count with GitHub actions?

GitHub Actions has a unique ID number for every workflow but the date + time is more useful.

Reply:
> Drawback: random updates from a PPA during one day

Is this a problem? If the user is installing from the beta PPA, I presume they want the latest code.

Reply:
It does not work to use just date + short commit hash, because the newer commit hash is not higher in any case.

Reply:
Right, the commit hash is just for our convenience. The date + time put the versions in order.

Reply:
How about using `printf "r%s.%s" "$(git rev-list --count HEAD)" "$(git rev-parse --short HEAD)"`?

Reply:
I do not want to keep the weird sequential revision number in anything. It is unnecessary. Let's just use commit date + time to sequentially version builds.

We cannot use such fancy technologies as printf on Windows. Fortunately Git has flexible date string formatting built in.

Reply:
Yes, this was just an example, we could use cmake or python as OS independent scripting language. If we use the date, can we leave out the dashes to make the name shorter?

Reply:
We could, but then it will be hard to understand what that big string of digits means.

Reply:
How about iso 8601? Something like 20201126T010850Z?

Reply:
Sure, we could separate the date and time with `T`. I don't think that's the same as ISO 8601 though.

```
> TZ=UTC git show --quiet --format=%cI
2020-11-25T19:02:14-06:00
```

Reply:
There are multiple represations defined in ISO8601. See section 4.3.2: http://dotat.at/tmp/ISO_8601-2004_E.pdf

Reply:
IMHO setting the package name doesn't belong into the CMakeLists.txt file. I don't want to let our build system depend on the CI env vars (which would necessarily happen if we want to include stuff like PR number and so on).

Sorry, I didn't realize that when commenting here.

Reply:
> I don't want to let our build system depend on the CI env vars (which would necessarily happen if we want to include stuff like PR number and so on).

I agree. I think it should be easy to locally reproduce builds that are bit identical to the binaries we publish simply by checking out the commit and building. Unfortunately `git rev-parse --abbrev-ref HEAD` does not work to get the Git branch name with the way GitHub Actions clones the repository so this has to be hacked around. If you have a better idea, please share.

Why shouldn't the be done in the CMakeLists.txt file?

Reply:
> I think it should be easy to locally reproduce builds that are bit identical to the binaries we publish simply by checking out the commit and building.

I agree. 

> Unfortunately `git rev-parse --abbrev-ref HEAD` does not work to get the Git branch name with the way GitHub Actions clones the repository so this has to be hacked around. If you have a better idea, please share.

Have you tried passing `fetch-depth: 0` to the checkout action? 

> Why shouldn't this be done in the CMakeLists.txt file?

I think we can run the git commands necessary to populate the about dialog in the CMakeLists.txt, but I don't think we need to set the package file name there. It's something the user/packager should decide. I think CPACK_PACKAGE_FILE_NAME just modifies the file name, so the package should still be bit-identical. 

For our CI, we should include the PR number in the file name like we did on Appveyor. Currently working on that.

Reply:
I'm keen on getting the git hash in the "About" menu as well.

> I agree. I think it should be easy to locally reproduce builds that are bit identical to the binaries we publish simply by checking out the commit and building. Unfortunately `git rev-parse --abbrev-ref HEAD` does not work to get the Git branch name with the way GitHub Actions clones the repository so this has to be hacked around. If you have a better idea, please share.

It's fairly common practice to reference git commits via this:
`$git describe --tags --always` 
which gives
`release-2.2.4-8697-g15867d988b`.
Then you can checkout that commit to a local branch with:
`git checkout release-2.2.4-8697-g15867d988b -b test_branch`. 

With git the branches are really just virtual as everything hangs off the same chain of hashes.

Reply:
> release-2.2.4-8697-g15867d988b

That doesn't work for us now because 2.2.4 is the last stable release, not the current beta. If we decide to use this technique we could retroactively make a 2.3.0-beta tag.

Reply:
I was not aware of `git describe`. Thank you for sharing that.

Reply:
I think we should use `git describe` instead of inventing our own format. If there is consensus to do this, then we need to push a tag to mark 2.3.0-beta. I think we should drop the `release-` prefix from our tags. For the package file names, those could simply be:

```
Mixxx-`git describe --tags --always`.FILE_EXTENSION
```

I think there would be no need to put the commit date and time in the file name, but that would still be nice in the About dialog.

Reply:
I think we should tag e60a749bc20412b938e923d39832ba110eb7db9b as `2.3.0-beta`.

Reply:
`git describe` will work for the Debian package naming for the PPA if we use `X.Y.Z~beta` for beta tags. We must use the `~` to separate `beta` from the version number:

```
> dpkg --compare-versions 2.3.0~beta-3852-g1234567 lt 2.3.0-1-g1234560 && echo true
true
> dpkg --compare-versions 2.3.0-beta-3852-g1234567 lt 2.3.0-1-g1234560 && echo true
```

Reply:
Nice